### ISSUE TITLE
Cannot resolve older assembly with newer version despite non-explicit version required

### ISSUE THREAD

**Issue by dapug** [Azure/azure-functions-host#2211]
If your project has (eg) assembly v2.4.x loaded but you reference a library that uses v2.2.x, You get a runtime error "Could not load file or assembly 'Blah, Version=2.2.0.0"

- Specifically, my Functions app is using WindowsAzure.Storage v8.6.0.0'
- I'm referencing a library (Azure.TableStorage.Repository) that happens to have a dependency on storage v8.4.0.0
- I verify in NuGet it specifically lists WindowsAzure.Storage (>= 8.4.0)  
- Note the **greater-than**, not just equal to. It is _not an explicit version reference in that library_!

#### Investigative information

- Microsoft.NET.Sdk.Functions 1.0.2
- Microsoft.Azure.WebJobs 2.1.0-beta1
- VS 2017 v15.5.0 Preview 4.0
- Running locally / emulated (not in Azure cloud)

#### Repro steps

1. Create a new Azure Functions project
2. Add NuGet WindowsAzure.Storage v8.6.0.0
3. Add NuGet Azure.TableStorage.Repository v1.0.6 (with >= v8.4.0 WindowsAzure.Storage dependency)
4. Create a TimerTrigger function (for convenience) that tries to insert a record into a table using this library
5. Debug -> Start New instance
-->

#### Expected behavior

The Azure Storage library should load v8.6.0 and use it without complaint of failure despite the library that is referencing a lesser version.

#### Actual behavior

Exception:  `Could not load file or assembly 'Microsoft.WindowsAzure.Storage, Version=8.4.0.0`

#### Known workarounds

None

#### Related information 

The exact same code (call to Azure table using this 3rd party lib) in a Unit Test app works perfectly fine, so it's not the 3rd party lib that has the issue.  Meaning, my .Net Core Unit Test app has Microsoft.WindowsAzure.Storage v8.6.0 loaded AND the Azure.TableStorage.Repository library and makes use of that library without issue, resolving the dependency appropriately. **It only chokes under a Functions app environment**.

**Comment by fabiocav** [comment]
We have a runtime enhancement about to land that will assist with these scenarios but currently this might be a bit challenging depending on the route you take.

Do you need to use 8.6.0? Or would 8.4.0 (the version that matches the transient dependency introduced by Azure.TableStorage.Repository) suffice? If you don't need 8.6.0, simply removing the explicit reference you have in your project should help.

If you need to use 8.6.0, you'd need to handle the assembly resolution. This is something I'd recommend against, unless you have no other alternative.

The feature I mentioned above will be landing at the beginning of the year.

**Comment by dapug** [comment]
v8.4.0 in this case is sufficient, so downgrading to "match" was ok for me.  I imagine that won't be the case for others, or for me going forward.  Kudos to the team for progress.  I look forward to more updates.

**Comment by christopheranderson** [comment]
@fabiocav - can you create an issue to track the work you're planning so we can reference that issue for these class of bugs?

**Comment by jeroenwo** [comment]
I also have the same issue, running Durable Functions. But downgrading to 8.4.0.0 did not work for me:
`[1/24/2018 06:01:56 AM] System.Private.CoreLib: Exception while executing function: RetrieveRedisMetrics. Nextens.StatusDashboards.Data.AzureStorage: Could not load file or assembly 'Microsoft.WindowsAzure.Storage, Version=8.4.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621). System.Private.CoreLib: Could not load file or assembly 'Microsoft.WindowsAzure.Storage, Version=8.4.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.`
- Using a project reference which uses 8.4.0
- Using an AF which references Microsoft.Azure.Sdk.Functions (v1.0.6)
- Which uses Microsoft.Azure.WebJobs (v3.0.0-beta3)
- Which uses WindowsAzure.Storage (v8.4.0)

- And using Microsoft.Azure.WebJobs.Extensions.DurableTask (v1.0.0-beta)
- Which uses DurableTask.AzureStorage (v1.0.0-beta)
- Which uses WindowsAzure.Storage (v8.4.0)

The AF itself does not reference WindowsAzure.Storage package (the result is the same if I do reference it).

If I remove any references to the project using WindowsAzure.Storage, clean and build, I can see the WindowsAzure.Storage (v8.4.0) in the bin directory.

Any ideas?

**Comment by jeroenwo** [comment]
Example:
```csharp
public static class RetrieveData
{
    [FunctionName("RetrieveData")]
    public static async Task RunOrchestrator([TimerTrigger("0 */1 * * * *")] TimerInfo timerInfo,
        [OrchestrationClient] DurableOrchestrationClient client, TraceWriter log)
    {

        // Replace "hello" with the name of your Durable Activity Function.
        await client.StartNewAsync("RetrieveEdgarData_Hello", "Tokyo");
        await client.StartNewAsync("RetrieveEdgarData_Hello", "Seattle");
        await client.StartNewAsync("RetrieveEdgarData_Hello", "London");

        var storageAccount = CloudStorageAccount.Parse(Environment.GetEnvironmentVariable("Values:AzureWebJobsStorage"));
    }

    [FunctionName("RetrieveDataData_Hello")]
    public static void SayHello([OrchestrationTrigger] DurableOrchestrationContext context, TraceWriter log)
    {
        var name = context.GetInput<string>();
        log.Info($"Saying hello to {name}.");
    }
}
```

**Comment by jeroenwo** [comment]
While a normal Function requires 8.4.0, a **Durable Function** requires 8.3.0. So the referenced project requires a reference to 8.3.0, and I have removed the AzureStorage reference from the function itself. And then it works..

**Comment by paulbatum** [comment]
@jeroenwo I would file a new issue for the durable functions specific problem
@dapug closing this as you were able to unblock and there are already issues tracking assembly resolution

**Comment by marcin-chwedczuk-meow** [comment]
@paulbatum "there are already issues tracking assembly resolution" - could you provide links to these issues here?

**Comment by paulbatum** [comment]
The main one where most of the discussion lies is: https://github.com/Azure/azure-functions-host/issues/992
